controllers/service_controller_unit_test.go

```diff
@@ -0,0 +1,253 @@
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+const (
+
+	// Don't using the const here just because if you change the const, the behaviour
+	// will change, and things will not work as expected, keep it separate.
+
+	TestKuadrantDiscoveryAnnotationOASConfigMap = "discovery.kuadrant.io/oas-configmap"
+	TestKuadrantDiscoveryAnnotationOASPath      = "discovery.kuadrant.io/oas-path"
+	TestKuadrantDiscoveryAnnotationOASNamePort  = "discovery.kuadrant.io/oas-name-port"
+
+	PetStoreOAS = `
+openapi: "3.0.0"
+info:
+	title: "Petstore Service"
+	version: "1.0.0"
+servers:
+	- url: https://petstore.swagger.io/v1
+paths:
+	/pets:
+		get:
+			operationId: "getPets"
+			responses:
+				405:
+					description: "invalid input"`
+)
+
+func getReconciler() *reconcilers.BaseReconciler {
+
+	s := scheme.Scheme
+	err := appsv1.AddToScheme(s)
+	if err != nil {
+		return nil
+	}
+
+	CatOASConfigMap := &v1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ConfigMap",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "cat-oas",
+			Namespace: "test",
+		},
+		Data: map[string]string{
+			"openapi.yaml": PetStoreOAS,
+		},
+	}
+
+	DogOASConfigMap := &v1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ConfigMap",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "dog-oas",
+			Namespace: "test",
+		},
+		Data: map[string]string{
+			"openapi-invalid.yaml": PetStoreOAS,
+		},
+	}
+	// Objects to track in the fake client.
+	objs := []runtime.Object{CatOASConfigMap, DogOASConfigMap}
+
+	// Create a fake client to mock API calls.
+	cl := fake.NewFakeClient(objs...)
+	clientAPIReader := fake.NewFakeClient(objs...)
+	recorder := record.NewFakeRecorder(10000)
+
+	baseReconciler := reconcilers.NewBaseReconciler(cl, s, clientAPIReader, LOGTEST, recorder)
+	return baseReconciler
+}
+
+func getServiceReconciler() *ServiceReconciler {
+	return &ServiceReconciler{getReconciler()}
+}
+
+func getSampleService() *v1.Service {
+	return &v1.Service{
+		TypeMeta: metav1.TypeMeta{APIVersion: "v1", Kind: "Service"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        "test",
+			Namespace:   "test",
+			Labels:      map[string]string{KuadrantDiscoveryLabel: "true"},
+			Annotations: map[string]string{},
+		},
+		Spec: v1.ServiceSpec{
+			Ports: []v1.ServicePort{
+				{Port: 8080, Name: "test", Protocol: v1.ProtocolTCP, TargetPort: intstr.FromInt(3000)},
+				{Port: 10000, Name: "openapi", Protocol: v1.ProtocolTCP, TargetPort: intstr.FromInt(3000)},
+			},
+			Selector: map[string]string{"svc": "cats"},
+		},
+	}
+}
+
+// TestOasDiscoveryNoAnnotation test that check if no annotation nothing
+// happens at all
+func TestOasDiscoveryNoAnnotation(t *testing.T) {
+	serviceReconciler := getServiceReconciler()
+	svc := getSampleService()
+	log := serviceReconciler.Logger().WithValues("test", "foo")
+	hasOas, result, err := serviceReconciler.IsOasDefined(context.Background(), svc, log)
```

Comment:
The testing approach is generally ok. It is testing main use cases, without annotations, with OAS in configmap and with OAS served by upstream. 

However it is being tested calling a `IsOasDefined` method that should, IMO, be private, hence non testeable. Another approach would be instantiate the reconciler and call `Reconcile()` method and then read current "fake cluster" status for expected results. We have been using this testing approach in other controllers:

```go
 endLoop := false                                                    
 for i := 0; i < 100 && !endLoop; i++ {                              
     res, err := r.Reconcile(req)                                    
     if err != nil {                                                 
         t.Fatal(err)                                                
     }                                                               
                                                                     
     endLoop = !res.Requeue                                          
 }                                                                   
                                                                     
 if !endLoop {                                                       
     t.Fatal("reconcile did not finish end of reconciliation as expected")
 }                                                                   
```

For the edge case where the endpoint is not available, the test can check the error returned by the reconcile and make sure the error matches. WDTY? it makes sense?

Reply:
The test of reconcile is used on some integration test, does not make much sense to me to have the unit test regarding that. I mean, OAS config function is quite simple, and has 100% coverage.

Reply:
Up to you. But `IsOasDefined` is being public just to be tested in the unittests and looks like it should be private. That is a bad smell that something does not fit well.